lib/fetch/index.js

```diff
@@ -122,130 +124,170 @@ class Fetch extends EE {
+      }
 
-    // 2. If response’s aborted flag is set, then:
-    if (response.aborted) {
-      // 1. Let deserializedError be the result of deserialize a serialized
-      //    abort reason given controller’s serialized abort reason and
-      //    relevantRealm.
+      try {
+        // 2. If response’s aborted flag is set, then:
+        if (response.aborted) {
+          // 1. Let deserializedError be the result of deserialize a serialized
+          //    abort reason given controller’s serialized abort reason and
+          //    relevantRealm.
 
-      // 2. Abort the fetch() call with p, request, responseObject, and
-      //    deserializedError.
+          // 2. Abort the fetch() call with p, request, responseObject, and
+          //    deserializedError.
 
-      abortFetch(p, request, responseObject, controller.serializedAbortReason)
-      return Promise.resolve()
-    }
+          abortFetch(p, request, responseObject, controller.serializedAbortReason, context)
+          return Promise.resolve()
+        }
 
-    // 3. If response is a network error, then reject p with a TypeError
-    // and terminate these substeps.
-    if (response.type === 'error') {
-      p.reject(new TypeError('fetch failed', { cause: response.error }))
-      return Promise.resolve()
-    }
+        // 3. If response is a network error, then reject p with a TypeError
+        // and terminate these substeps.
+        if (response.type === 'error') {
+          context.error = response.error
+          subscribed('fetchError')
+          context.error = null
+          p.reject(new TypeError('fetch failed', { cause: response.error }))
+          return Promise.resolve()
+        }
 
-    // 4. Set responseObject to the result of creating a Response object,
-    // given response, "immutable", and relevantRealm.
-    responseObject = fromInnerResponse(response, 'immutable', relevantRealm)
+        // 4. Set responseObject to the result of creating a Response object,
+        // given response, "immutable", and relevantRealm.
+        responseObject = fromInnerResponse(response, 'immutable', relevantRealm)
 
-    // 5. Resolve p with responseObject.
-    p.resolve(responseObject)
-  }
+        context.result = responseObject[kState]
 
-  controller = fetching({
-    request,
-    processResponseEndOfBody: handleFetchDone,
-    processResponse,
-    dispatcher: init?.dispatcher ?? getGlobalDispatcher() // undici
-  })
+      // 5. Resolve p with responseObject.
+      } finally {
+        callAsyncStartEnd()
+        p.resolve(responseObject)
+      }
+    }
 
+    controller = fetching({
+      request,
+      processResponseEndOfBody: handleFetchDone,
+      processResponse,
+      dispatcher: init?.dispatcher ?? getGlobalDispatcher() // undici
+    })
+  } finally {
+    subscribed('end')
+  }
   // 14. Return p.
-  return p.promise
+  return p.promise.then(() => {
+    callAsyncStartEnd()
+    context.result = null
+  })
```

Comment:
Please avoid allocating an additional promise here in case the diagnostic channels are not enabled. Promises are not free.

Reply:
@mcollina are you referring to the .then()? the p.promise was there before and I just added .then off of it in order to publish the events

Reply:
Yes! You can avoid it if the channel does not have listeners.

Reply:
added an if block to check for subscribers before calling .then